README
PDF-Table version 0.9.4
=====================
SOME NOTES

This module is intended for table generation using PDF::API2
The current version is RC1 and I will apreciate any feedback.
Developed and tested on i586 Linux SuSE 10.0 and perl, v5.8.7 built for i586-linux-thread-multi

CHANGES

Since version 0.02 there are many changes. 
See the ChangeLog file or make a diff from the tools menu in CPAN

CONTACTS 

See http://search.cpan.org/~jbazik/
See http://search.cpan.org/~omega/

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  PDF::API2

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2006 by Daemmon Hughes

Extended by Desislav Kamenov since version 0.02

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.7 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension PDF::Table.

Version 0.9.4 - 17.Oct.2011
----------------------------
Don't bump row count when row spans a page.  Closes #67401.

Fixed bug: don't call text_block if record width is exactly cell width.

Fixed bug in text_block: bottom is (top - height), not (baseline - height).

Example script typo fixed.  Closes #55133.

Don't set up strokes when border is 0.  Apparently breaks Acrobat.
Closes #40705 and #55137.

Changed print warnings to carp.  Closes #38007.

Added center justification for columns and cells.  Tests, too.
Closes #37924.

Fixed infinite loop caused by uninitialized $2.  Thanks to Petri Kaurinkoski.
Closes #34017.

Fixed perl warning if $col_prop->{min_w} is undef.  Thanks to
Matthias Pitzl (NEPOS).  Closes #30720.

Fixed default header background color bug, closes #28542.

Added mock PDF::API2 packages to test script and added tests.

Improve column width calculations so wider content gets wider columns.
Added column width test.  Thanks to Barrie Slaymaker (RBS).  Closes #26158.


Version 0.9.3 - 27.12.2006
----------------------------
Added backward compatibility and now again params with '-' front are supported.
  Both params with '-'and without are valid and can be even mixed together.
  I excuse for changing the interface before.

Added cell_props parameter wich can define properties for a single cell as specified by user.
Thanks for cell_props patch to Bodo Bauer.

Removed Version requirement of Perl 5.8.7 and only version ot 5.6.0 is required.
Thanks to Jonas Pasche for reporting it.

Updated examples/sample1.pl file to try the capabilities of the module immidiately after download.

Documentation Update.

Fixed bug with interpreting \n as special char but not as newline.
Thanks to Jonas Pasche for reporting it.
	(reminder: text with \n must be enclosed in double quotes i.e. "text\nnewline" )

Fixed documentation inacuracy for text_block(). It is called as $pdftable->text_block() of course.
Thanks to Jonas Pasche for reporting it.

Fixed bug with first_line overlaping with former paragraph last line.
Thanks to Jonas Pasche for reporting it and contributing a patch.


Version 0.9.2 - 02.Oct.2006
----------------------------
Fixed bug with 'col_props->bg_color' overiding 'header_props->bg_color'
Thanks to Giulio Di Giuliomaria

Fixed some documentation inacuracys.
Added some new documentation.

'column_props->width' parameter changed to 'min_w'

Added example script to instantly test the module when downloaded.
See 'examples' into the tar.gz archive.

Version 0.9.1 - 18.Sep.2006
----------------------------
Removed Debug Message "Created New Page"

Fixed bug with not working header row bg_color parameter

Thanks to Bodo Bauer for submiting the patch.


Version 0.9 - 14.Sep.2006
----------------------------
Fixed 'fontsize' bug at line 234 to font_size and parameter is working correctly

Changed all parameters not to start with '-'
So now '-start_h' becomes only 'start_h'

Changed default font encoding form latin1 to UTF8

Added Default Values for next_h next_y parameters

Added Minimum value check for x y start_h start_y

Optimization for min row height calculations

Added 'new_page_func parameter'. This is should be a reference to function which is called every time when a new page needs to be created so the table can span.See online manual for more info how to use it.

Added Check if any text in sub text_block before spliting the parameter.

Fixed bug with infinite loop when spanning to a new page.

Splitting long words to parts because of breaking table cels

Parameter in text_block() 'parspace'  space before block start (paragraph).
So the first line will start and height $ybase $parspace. So is every next paragraph into the block.

Undocumented Parameter in text_block() 'indent'    indent of every line
Undocumented Parameter in text_block() 'flindent'  indent of first line
Undocumented Parameter in text_block() 'fpindent'  indent of first paragraph

The Above parameters are excluding each other in the following order: parspace,flindent,fpindent,indent 
The one that is found first and has valid value is beeing used

Added Header Row support with repeat on every new page and different layout than the other rows

Parameter 'w' for table is not optional any more (could change)

Changed table width calculation and now module tries to use as much space as it is given but not more than it.

Default values for many options

Better source formatting

Fixed incorect y position returned from text_block()

Others that i dont remeber.


examples/pdf_table_test.pl
#!/usr/bin/perl
use warnings;
use strict;
use diagnostics;
#Please use TABSTOP=4 for best view
 use PDF::API2;
 use PDF::Table;

 my $pdftable = new PDF::Table;
 my $pdf = new PDF::API2(-file => "table_of_lorem.pdf");
 my $page = $pdf->page();
 $pdf->mediabox('A4');
 # A4 as defined by PDF::API2 is h=842 w=545 for portrait
	
 # some data to layout
 my $some_data =[
        ['Header',
        'Row',
        'Test'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
 ];

 # build the table layout
 $pdftable->table(
         # required params
         $pdf,
         $page,
         $some_data,
         x => 50,
         w => 495,
         start_y => 792,
         next_y  => 700,
         start_h => 400,
         next_h  => 500,
         # some optional params
         padding => 5,
         padding_right => 10,
         background_color_odd  => 'lightblue',
         background_color_even => "#EEEEAA", #cell background color for even rows
         header_props          => { 
             bg_color   => "#F0AAAA",
             font       => $pdf->corefont("Helvetica", -encoding => "utf8"),
             font_size  => 14,
             font_color => "#006600",
             repeat     => 1,
         },
         column_props => [
            {},
            {
                min_w => 250,
                justify => "right",
                font => $pdf->corefont("Times", -encoding => "latin1"),
                font_size => 14,
                font_color=> 'white',
                background_color => '#262633',
            },
         ],
  );
$pdf->saveas();


examples/sample1.pl
#!/usr/bin/perl
use warnings;
use strict;
use diagnostics;
#Please use TABSTOP=4 for best view
 use PDF::API2;
 use Table;

 my $pdftable = new PDF::Table;
 my $pdf = new PDF::API2(-file => "table_of_lorem.pdf");
 my $page = $pdf->page();
 $pdf->mediabox('A4');
 # A4 as defined by PDF::API2 is h=842 w=545 for portrait
	
 # some data to layout
 my $some_data =[
        ['Header',
        'Row',
        'Test'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        "i ne razbiram DESI\ngorniq \nezik",
        "zatova reshih
		da dobavq
		edin ili dva
		novi reda"],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
        ['1 Lorem ipsum dolor',
        'Donec odio neque, faucibus vel',
        'consequat quis, tincidunt vel, felis.'],
        ['Nulla euismod sem eget neque.',
        'Donec odio neque',
        'Sed eu velit.'],
        ['Az sym bulgarin',
        'i ne razbiram gorniq ezik',
        'zatova reshih'],
        ['da dobavq edin dva reda',
        'v tozi primer',
        'na bulgarski ezik s latinica'],
 ];

 # build the table layout
 $pdftable->table(
         # required params
         $pdf,
         $page,
         $some_data,
         x => 50,
         -w => 495,
         start_y => 792,
         next_y  => 700,
         -start_h => 400,
         next_h  => 500,
         # some optional params
         -padding => 3,
         padding_right => 10,
         background_color_odd  => 'lightblue',
         background_color_even => "#EEEEAA", #cell background color for even rows
         header_props          => { 
             bg_color   => "#F0AAAA",
             font       => $pdf->corefont("Helvetica", -encoding => "utf8"),
             font_size  => 14,
             font_color => "#006600",
             repeat     => 1,
         },
         column_props => [
            {},
            {
                min_w => 250,
                justify => "right",
                font => $pdf->corefont("Times", -encoding => "latin1"),
                font_size => 14,
                font_color=> 'white',
                background_color => '#8CA6C5',
            },
         ],
         cell_props => [
           [#This is the first(header) row of the table and here wins header_props
             {
               background_color => '#000000',
               font_color       => 'blue',
             },
             # etc.
           ],
           [#Row 2 
             {#Row 2 cell 1
               background_color => '#000000',
               font_color       => 'white',
             },
             {#Row 2 cell 2
               background_color => '#AAAA00',
               font_color       => 'red',
             },
             {#Row 2 cell 3
               background_color => '#FFFFFF',
               font_color       => 'green',
             },
             # etc.
           ],
           [#Row 3
             {#Row 3 cell 1
               background_color => '#AAAAAA',
               font_color       => 'blue',
             },
             # etc.
           ],
           # etc.
         ],
  );
$pdf->saveas();
